### PR TITLE
Enhance data driven notification

### PR DESCRIPTION
**Pull request by vertein**
Enhanced data notifications to optionally specify the notification message itself.

Also adds in unit tests that were removed by earlier commits.

----

Contributor License Agreement adherence:

- [x] This Contribution is under the terms of Individual [Contributor License Agreements][] (and also Corporate Contributor License Agreements to the extent applicable) appearing in the [Apereo CLA roster][].

[Apereo CLA roster]: http://licensing.apereo.org/completed-clas
[Contributor License Agreements]: https://www.apereo.org/licensing/agreements

### CODE REVIEW

File: docs/messaging-implementation.md [hunk 1]
```diff
@@ -66,6 +71,8 @@ Widget messaging is based on JSON input configured in a [widget's configuration]
     You would use this feature if you want to only show the message if the specific user has data. For example: Only show user if they have a certain document.
   - **dataObject**: *(optional)* Will only be looked at if `dataUrl` is present, otherwise ignored. Used as an optional further refinement from dataUrl, if you want the notification to show only if the specific object is in the data.
   - **dataArrayFIlter**: Will only be looked at if `dataUrl` is present, otherwise ignored. Used as an optional further refinement from dataUrl. If your object return is an array, you can filter on the array. Does support multiple filtering criteria as shown in the example. If used in conjunction with `dataObject`, will filter to `dataObject` first.  [AngularJS array filtering documentation] (https://docs.angularjs.org/api/ng/filter/filter)
```
Comment (ChristianMurphy) [inline docs/messaging-implementation.md]: :art: unrelated to these changes but I just spotted `dataArrayFIlter` -> `dataArrayFilter`

File: docs/messaging-implementation.md [hunk 2]
```diff
@@ -66,6 +71,8 @@ Widget messaging is based on JSON input configured in a [widget's configuration]
...
+  - **dataMessageTitle** Will be used if dataUrl is specified. Used to set the title of the message from the data response from `dataUrl`.  Expects an array for where to find the title in the data response from `dataUrl`.
```
Comment (ChristianMurphy) [inline docs/messaging-implementation.md]: :question: would it be easier if there was a `data` block in the json with `Object`, `ArrayFilter`, `MessageTitle`, and `MessageMoreInfoUrl` as child properties?

Comment (vertein) [reply docs/messaging-implementation.md]: Good insight.  Yeah, probably.

Comment (vertein) [reply docs/messaging-implementation.md]: Moved to a new `data` block.

File: docs/messaging-implementation.md (hunk 1 above)
Comment (vertein) [reply docs/messaging-implementation.md]: unrelated, but was a quick win.  Updated.

File: docs/messaging-implementation.md [hunk 3]
```diff
@@ -60,12 +67,15 @@ Widget messaging is based on JSON input configured in a [widget's configuration]
...
   - **dataObject**: *(optional)* Will only be looked at if `dataUrl` is present, otherwise ignored. Used as an optional further refinement from dataUrl, if you want the notification to show only if the specific object is in the data.
-  - **dataArrayFIlter**: Will only be looked at if `dataUrl` is present, otherwise ignored. Used as an optional further refinement from dataUrl. If your object return is an array, you can filter on the array. Does support multiple filtering criteria as shown in the example. If used in conjunction with `dataObject`, will filter to `dataObject` first.  [AngularJS array filtering documentation] (https://docs.angularjs.org/api/ng/filter/filter)
+  - **dataArrayFilter**: Will only be looked at if `dataUrl` is present, otherwise ignored. Used as an optional further refinement from dataUrl. If your object return is an array, you can filter on the array. Does support multiple filtering criteria as shown in the example. If used in conjunction with `dataObject`, will filter to `dataObject` first.  [AngularJS array filtering documentation] (https://docs.angularjs.org/api/ng/filter/filter)
```
Comment (ChristianMurphy) [inline docs/messaging-implementation.md]: :thinking: :question: with the `data` object wrapping, is the `data` prefix on the property still wanted?

Comment (vertein) [reply docs/messaging-implementation.md]: Yes.  I think the names aren't that great to begin (I created them, so I don't feel too bad saying that) 
 and without the data prefix, they might even be worse.  We need some new names here.

Comment (ChristianMurphy) [reply docs/messaging-implementation.md]: Relevant, great quote

> There are only two hard things in Computer Science: cache invalidation and naming things.
> -- Phil Karlton

source: https://www.martinfowler.com/bliki/TwoHardThings.html

**Pull request merged**